packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -649,223 +600,326 @@ export function getValidatorApi(
+    feeRecipient,
+    builderSelection,
+    strictFeeRecipientCheck,
+    slot,
+    randaoReveal,
+    graffiti,
+    builderBoostFactor,
+    forceBlockType,
+  }: routes.validator.ExtraProduceBlockOpts & {
+    slot: Slot;
+    randaoReveal: BLSSignature;
+    graffiti?: string;
+    // TODO deneb: skip randao verification
+    skipRandaoVerification?: boolean;
+    builderBoostFactor?: bigint;
+    forceBlockType?: BlockType;
+  }): Promise<ProduceFullOrBlindedBlockOrContentsRes> {
+    notWhileSyncing();
+    await waitForSlot(slot); // Must never request for a future slot > currentSlot
 
-      if (executionPayloadSource === ProducedBlockSource.engine) {
-        return {
-          ...engine.value,
-          executionPayloadBlinded: false,
-          executionPayloadSource,
-        };
+    const parentBlockRoot = fromHex(chain.getProposerHead(slot).blockRoot);
+    notOnOutOfRangeData(parentBlockRoot);
+    const ctx = normalizeBlockProductionOptions({slot, builderBoostFactor, builderSelection, forceBlockType});
+
+    const blockAttributes: BlockAttributes = {
+      feeRecipient,
+      randaoReveal,
+      graffiti: toGraffitiBuffer(
+        graffiti ?? getDefaultGraffiti(getLodestarClientVersion(opts), chain.executionEngine.clientVersion, opts)
+      ),
+      parentBlockRoot,
+      slot,
+    };
+
+    const loggerContext = {
+      slot,
+      fork: ctx.fork,
+      builderSelection,
+      strictFeeRecipientCheck,
+      // winston logger doesn't like bigint
+      builderBoostFactor: `${builderBoostFactor}`,
+    };
+
+    // Starting the process of block production
+    const blockProductionTimer = metrics?.blockProductionTime.startTimer();
+    const {block, executionPayloadValue, consensusBlockValue, executionPayloadSource, executionPayloadBlinded} =
+      await produceAssembledBlock(blockAttributes, ctx, loggerContext);
+    blockProductionTimer?.({source: executionPayloadSource});
+
+    logger.verbose("Produced block", {
+      slot: block.slot,
+      executionPayloadValue,
+      consensusBlockValue,
+      root: toRootHex(config.getForkTypes(block.slot).BeaconBlock.hashTreeRoot(block)),
+    });
+
+    metrics?.blockProductionNumAggregated.observe({source: executionPayloadSource}, block.body.attestations.length);
+    metrics?.blockProductionExecutionPayloadValue.observe(
+      {source: executionPayloadSource},
+      Number(formatWeiToEth(executionPayloadValue))
+    );
+
+    const version = config.getForkName(block.slot);
+    if (strictFeeRecipientCheck && feeRecipient && isForkPostBellatrix(version)) {
+      const blockFeeRecipient = toHex((block as bellatrix.BeaconBlock).body.executionPayload.feeRecipient);
+      if (blockFeeRecipient !== feeRecipient) {
+        throw Error(`Invalid feeRecipient set in engine block expected=${feeRecipient} actual=${blockFeeRecipient}`);
+      }
+    }
+
+    if (chain.opts.persistProducedBlocks) {
+      void chain.persistBlock(block, "produced_engine_block");
```

Comment:
this code and the above belongs to produce engine block flow which seems to be misplaced in the common code between `builder + engine` here
need an `if (executionPayloadSource === ProducedBlockSource.engine)`?